python/src/iceberg/schema.py

```diff
@@ -0,0 +1,67 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from .types import StructType, Type, NestedField
+
+
+class Schema(object):
+    alias_to_id: dict = None
+    id_to_field = {}
+    name_to_id: dict = None
+    lowercase_name_to_id: dict = None
```

Comment:
IMO, those 4 fields can be included within `__init__`.
Also, if they are not needed in this change, we might consider to add them in the future when we actually need them

Reply:
I think these are simply class variables that might be used for methods. They don't need to be part of the `__init__` since they aren't initialized externally. 
Let me fix the PR with the rest of the changes and see if we need the variables for this iteration of change. I'll remove them if not needed.

Reply:
It seems that here is declaring mutable variables as class attributes. This can be problematic because class attributes are mutable and persist across the same python session. i.e.

```python
class Spec:
    id: str = None
    name: str = None


spec1 = Spec()
spec2 = Spec()

assert(spec1.id is None)

# change class variable from an instance
spec2.__class__.id = "spec2_id"
assert(spec1.id == "spec2_id")
assert(Spec.id == "spec2_id")

spec3 = Spec() 
assert(spec3.id == "spec2_id")

Spec.id = "class_spec_id"
assert(spec1.id == "class_spec_id")
assert(spec2.id == "class_spec_id")
assert(spec3.id == "class_spec_id")
```

In the above example, an instance can change the class variable and it affects all other current and future instances. This can lead to surprising behavior.

To have instance scope only variables, we can put them inside the init without exposing them to the signature. i.e.

```python
   def __init__(self, struct: StructType, schema_id: int, identifier_field_ids: [int]):
        self._struct = struct
        self._schema_id = schema_id
        self._identifier_field_ids = identifier_field_ids
        # not initialized by the user 
        self.alias_to_id: dict = None
        self.id_to_field = {}
        self.name_to_id: dict = None
        self.lowercase_name_to_id: dict = None
```

Reply:
Agreed. This is the approach that makes sense. Fixing in the upcoming commits.